// @atlaskit/editor-core/dist/esm/plugins/extension/ui/Extension/ExtensionComponent.js
import { __extends } from "tslib";
import React from 'react';
import { Component } from 'react';
import { selectParentNodeOfType, findSelectedNodeOfType, } from 'prosemirror-utils';
import { getExtensionRenderer, getNodeRenderer, } from '@atlaskit/editor-common';
import { setNodeSelection } from '../../../../utils';
import Extension from './Extension';
import InlineExtension from './InlineExtension';
var ExtensionComponent = /** @class */ (function (_super) {
    __extends(ExtensionComponent, _super);
    function ExtensionComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {};
        _this.mounted = false;
        _this.setStateFromPromise = function (stateKey, promise) {
            promise &&
                promise.then(function (p) {
                    var _a;
                    if (!_this.mounted) {
                        return;
                    }
                    _this.setState((_a = {},
                        _a[stateKey] = p,
                        _a));
                });
        };
        _this.handleSelectExtension = function (hasBody) {
            var _a = _this.props.editorView, state = _a.state, _b = _a.state, selection = _b.selection, schema = _b.schema, dispatch = _a.dispatch;
            var tr = state.tr;
            if (hasBody) {
                tr = selectParentNodeOfType([schema.nodes.bodiedExtension])(state.tr);
                dispatch(tr);
            }
            else if (!findSelectedNodeOfType([
                schema.nodes.inlineExtension,
                schema.nodes.extension,
                schema.nodes.bodiedExtension,
            ])(selection)) {
                setNodeSelection(_this.props.editorView, selection.$from.pos - 1);
            }
        };
        _this.handleExtension = function (node) {
            var _a = _this.props, extensionHandlers = _a.extensionHandlers, editorView = _a.editorView;
            var _b = node.attrs, extensionType = _b.extensionType, extensionKey = _b.extensionKey, parameters = _b.parameters, text = _b.text;
            var isBodiedExtension = node.type.name === 'bodiedExtension';
            if (isBodiedExtension) {
                return;
            }
            var extensionParams = {
                type: node.type.name,
                extensionType: extensionType,
                extensionKey: extensionKey,
                parameters: parameters,
                content: text,
            };
            var result;
            if (extensionHandlers && extensionHandlers[extensionType]) {
                var render = getExtensionRenderer(extensionHandlers[extensionType]);
                result = render(extensionParams, editorView.state.doc);
            }
            if (!result) {
                var extensionHandlerFromProvider = _this.state.extensionProvider &&
                    getNodeRenderer(_this.state.extensionProvider, extensionType, extensionKey);
                if (extensionHandlerFromProvider) {
                    var NodeRenderer = extensionHandlerFromProvider;
                    return React.createElement(NodeRenderer, { extensionParams: extensionParams });
                }
            }
            return result;
        };
        return _this;
    }
    ExtensionComponent.prototype.UNSAFE_componentWillMount = function () {
        this.mounted = true;
    };
    ExtensionComponent.prototype.componentDidMount = function () {
        var extensionProvider = this.props.extensionProvider;
        if (extensionProvider) {
            this.setStateFromPromise('extensionProvider', extensionProvider);
        }
    };
    ExtensionComponent.prototype.componentWillUnmount = function () {
        this.mounted = false;
    };
    ExtensionComponent.prototype.UNSAFE_componentWillReceiveProps = function (nextProps) {
        var extensionProvider = nextProps.extensionProvider;
        if (extensionProvider &&
            this.props.extensionProvider !== extensionProvider) {
            this.setStateFromPromise('extensionProvider', extensionProvider);
        }
    };
    ExtensionComponent.prototype.render = function () {
        var _a = this.props, node = _a.node, handleContentDOMRef = _a.handleContentDOMRef, editorView = _a.editorView;
        var extensionHandlerResult = this.tryExtensionHandler();
        switch (node.type.name) {
            case 'extension':
            case 'bodiedExtension':
                return (React.createElement(Extension, { node: node, extensionProvider: this.state.extensionProvider, handleContentDOMRef: handleContentDOMRef, onSelectExtension: this.handleSelectExtension, view: editorView }, extensionHandlerResult));
            case 'inlineExtension':
                return (React.createElement(InlineExtension, { node: node }, extensionHandlerResult));
            default:
                return null;
        }
    };
    ExtensionComponent.prototype.tryExtensionHandler = function () {
        var node = this.props.node;
        try {
            var extensionContent = this.handleExtension(node);
            if (extensionContent && React.isValidElement(extensionContent)) {
                return extensionContent;
            }
        }
        catch (e) {
            // eslint-disable-next-line no-console
            console.error('Provided extension handler has thrown an error\n', e);
            /** We don't want this error to block renderer */
            /** We keep rendering the default content */
        }
        return null;
    };
    return ExtensionComponent;
}(Component));
export default ExtensionComponent;
