// @atlaskit/editor-core/dist/esm/plugins/width/index.js
import { __assign } from "tslib";
import { Plugin, PluginKey } from 'prosemirror-state';
export var pluginKey = new PluginKey('widthPlugin');
export function createPlugin(dispatch) {
    return new Plugin({
        key: pluginKey,
        state: {
            init: function () { return ({
                width: document.body.offsetWidth,
            }); },
            apply: function (tr, pluginState) {
                var meta = tr.getMeta(pluginKey);
                if (!meta) {
                    return pluginState;
                }
                var newPluginState = __assign(__assign({}, pluginState), meta);
                if (newPluginState &&
                    (pluginState.width !== newPluginState.width ||
                        pluginState.lineLength !== newPluginState.lineLength)) {
                    dispatch(pluginKey, newPluginState);
                    return newPluginState;
                }
                return pluginState;
            },
        },
    });
}
var widthPlugin = function () { return ({
    name: 'width',
    pmPlugins: function () { return [
        {
            name: 'width',
            plugin: function (_a) {
                var dispatch = _a.dispatch;
                return createPlugin(dispatch);
            },
        },
    ]; },
    // do this early here, otherwise we have to wait for WidthEmitter to debounce
    // which causes anything dependent on lineLength to jump around
    contentComponent: function (_a) {
        var editorView = _a.editorView, containerElement = _a.containerElement;
        var pmDom = containerElement
            ? containerElement.querySelector('.ProseMirror')
            : undefined;
        var newState = {
            lineLength: pmDom ? pmDom.clientWidth : undefined,
        };
        if (containerElement) {
            newState.width = containerElement.offsetWidth;
        }
        var tr = editorView.state.tr.setMeta(pluginKey, newState);
        editorView.dispatch(tr);
        return null;
    },
}); };
export default widthPlugin;
