// @atlassian/connect-module-core/dist/es6/modules/dialog/DialogModule.js
import * as tslib_1 from "tslib";
import { BaseModule } from '../base/Module';
import { DialogSimpleXdmDefinitionBuilder } from './DialogSimpleXdmDefinitionBuilder';
export var DialogModuleDefinitions;
(function (DialogModuleDefinitions) {
    DialogModuleDefinitions.DIALOG_MODULE_NAME = 'dialog';
    var Props = /** @class */ (function () {
        function Props(adaptor, dialogProvider) {
            this.adaptor = adaptor;
            this.dialogProvider = dialogProvider;
        }
        return Props;
    }());
    DialogModuleDefinitions.Props = Props;
    var State = /** @class */ (function () {
        function State(dialogProvider) {
            this.enabled = true;
            this.registered = false;
            this.dialogProvider = dialogProvider;
            this.simpleXdmDefinitionBuilder = new DialogSimpleXdmDefinitionBuilder(dialogProvider);
        }
        return State;
    }());
    DialogModuleDefinitions.State = State;
})(DialogModuleDefinitions || (DialogModuleDefinitions = {}));
/**
 * This module provides dialog related functionality.
 */
var DialogModule = /** @class */ (function (_super) {
    tslib_1.__extends(DialogModule, _super);
    function DialogModule(props) {
        return _super.call(this, props) || this;
    }
    DialogModule.prototype.componentWillMount = function () {
        if (!this.props.dialogProvider) {
            throw new Error('this.props.dialogProvider is not defined');
        }
        this.setState(new DialogModuleDefinitions.State(this.props.dialogProvider));
    };
    DialogModule.prototype.getModuleRegistrationName = function () {
        return DialogModuleDefinitions.DIALOG_MODULE_NAME;
    };
    DialogModule.prototype.getSimpleXdmDefinitionBuilder = function () {
        return this.state.simpleXdmDefinitionBuilder;
    };
    DialogModule.prototype.getProvider = function () {
        return this.props.dialogProvider;
    };
    DialogModule.prototype.render = function () {
        return null;
    };
    return DialogModule;
}(BaseModule));
export { DialogModule };
