// @atlaskit/editor-core/dist/esm/plugins/annotation/commands/index.js
import { INLINE_COMMENT } from '@atlaskit/adf-schema';
import { createCommand } from '../pm-plugins/inline-comment';
export var setInlineCommentState = function (newState) {
    return createCommand({
        type: 'SET_INLINE_COMMENT_STATE',
        data: newState,
    });
};
export var removeInlineCommentNearSelection = function (id) { return function (state, dispatch) {
    var tr = state.tr, $from = state.selection.$from;
    var annotationMarkType = state.schema.marks.annotation;
    var hasAnnotation = $from
        .marks()
        .some(function (mark) { return mark.type === annotationMarkType; });
    if (!hasAnnotation) {
        return false;
    }
    // just remove entire mark from around the node
    tr.removeMark($from.start(), $from.end(), annotationMarkType.create({
        id: id,
        type: INLINE_COMMENT,
    }));
    if (dispatch) {
        dispatch(tr);
    }
    return true;
}; };
export var resolveInlineComment = function (id) {
    return createCommand({ type: 'INLINE_COMMENT_RESOLVE', data: { id: id } });
};
