// @atlaskit/editor-core/dist/esm/plugins/base/pm-plugins/frozen-editor.js
import { Plugin } from 'prosemirror-state';
import { ACTION, ACTION_SUBJECT, EVENT_TYPE, } from '../../analytics';
import { isPerformanceObserverAvailable, isPerformanceAPIAvailable, } from '@atlaskit/editor-common';
var FREEZE_CHECK_TIME = 600;
var SLOW_INPUT_TIME = 300;
var DEFAULT_KEYSTROKE_SAMPLING_LIMIT = 100;
var dispatchLongTaskEvent = function (dispatchAnalyticsEvent, view, time) {
    var state = view.state;
    return dispatchAnalyticsEvent({
        action: ACTION.BROWSER_FREEZE,
        actionSubject: ACTION_SUBJECT.EDITOR,
        attributes: {
            freezeTime: time,
            nodeSize: state.doc.nodeSize,
        },
        eventType: EVENT_TYPE.OPERATIONAL,
    });
};
export default (function (dispatchAnalyticsEvent, inputSamplingLimit) {
    var keystrokeCount = 0;
    var samplingLimit = typeof inputSamplingLimit === 'number'
        ? inputSamplingLimit
        : DEFAULT_KEYSTROKE_SAMPLING_LIMIT;
    return new Plugin({
        props: isPerformanceAPIAvailable()
            ? {
                handleTextInput: function (view, from, to, text) {
                    var state = view.state;
                    var now = performance.now();
                    requestAnimationFrame(function () {
                        var diff = performance.now() - now;
                        if (samplingLimit && ++keystrokeCount === samplingLimit) {
                            keystrokeCount = 0;
                            dispatchAnalyticsEvent({
                                action: ACTION.INPUT_PERF_SAMPLING,
                                actionSubject: ACTION_SUBJECT.EDITOR,
                                attributes: {
                                    time: diff,
                                    nodeSize: state.doc.nodeSize,
                                },
                                eventType: EVENT_TYPE.OPERATIONAL,
                            });
                        }
                        if (diff > SLOW_INPUT_TIME) {
                            dispatchAnalyticsEvent({
                                action: ACTION.SLOW_INPUT,
                                actionSubject: ACTION_SUBJECT.EDITOR,
                                attributes: {
                                    time: diff,
                                    nodeSize: state.doc.nodeSize,
                                },
                                eventType: EVENT_TYPE.OPERATIONAL,
                            });
                        }
                    });
                    return false;
                },
            }
            : undefined,
        view: function (view) {
            if (!isPerformanceObserverAvailable()) {
                return {};
            }
            var observer;
            try {
                var observer_1 = new PerformanceObserver(function (list) {
                    var perfEntries = list.getEntries();
                    for (var i = 0; i < perfEntries.length; i++) {
                        var duration = perfEntries[i].duration;
                        if (duration > FREEZE_CHECK_TIME) {
                            dispatchLongTaskEvent(dispatchAnalyticsEvent, view, duration);
                        }
                    }
                });
                // register observer for long task notifications
                observer_1.observe({ entryTypes: ['longtask'] });
            }
            catch (e) { }
            return {
                destroy: function () {
                    if (observer) {
                        observer.disconnect();
                    }
                },
            };
        },
    });
});
