// @atlassian/connect-module-core/dist/es6/modules/message/MessageModule.js
import * as tslib_1 from "tslib";
import { BaseModule } from '../base/Module';
import { MessageSimpleXdmDefinitionBuilder } from './MessageSimpleXdmDefinitionBuilder';
export var MessageModuleDefinitions;
(function (MessageModuleDefinitions) {
    MessageModuleDefinitions.MESSAGE_MODULE_NAME = 'messages';
    var Props = /** @class */ (function () {
        function Props(adaptor, messageProvider) {
            this.adaptor = adaptor;
            this.messageProvider = messageProvider;
        }
        return Props;
    }());
    MessageModuleDefinitions.Props = Props;
    var State = /** @class */ (function () {
        function State(messageProvider) {
            this.enabled = true;
            this.registered = false;
            this.messageProvider = messageProvider;
            this.simpleXdmDefinitionBuilder = new MessageSimpleXdmDefinitionBuilder(messageProvider);
        }
        return State;
    }());
    MessageModuleDefinitions.State = State;
})(MessageModuleDefinitions || (MessageModuleDefinitions = {}));
/**
 * This module provides message related functionality.
 */
var MessageModule = /** @class */ (function (_super) {
    tslib_1.__extends(MessageModule, _super);
    function MessageModule(props) {
        return _super.call(this, props) || this;
    }
    MessageModule.prototype.componentWillMount = function () {
        if (!this.props.messageProvider) {
            throw new Error('this.props.messageProvider is not defined');
        }
        this.setState(new MessageModuleDefinitions.State(this.props.messageProvider));
    };
    MessageModule.prototype.getModuleRegistrationName = function () {
        return MessageModuleDefinitions.MESSAGE_MODULE_NAME;
    };
    MessageModule.prototype.getSimpleXdmDefinitionBuilder = function () {
        return this.state.simpleXdmDefinitionBuilder;
    };
    MessageModule.prototype.getProvider = function () {
        return this.props.messageProvider;
    };
    MessageModule.prototype.render = function () {
        return null;
    };
    return MessageModule;
}(BaseModule));
export { MessageModule };
