// @atlaskit/editor-core/dist/esm/plugins/text-color/index.js
import React from 'react';
import { textColor } from '@atlaskit/adf-schema';
import WithPluginState from '../../ui/WithPluginState';
import { pluginKey as textColorPluginKey, createPlugin, } from './pm-plugins/main';
import ToolbarTextColor from './ui/ToolbarTextColor';
var pluginConfig = function (textColorConfig) {
    if (!textColorConfig || typeof textColorConfig === 'boolean') {
        return undefined;
    }
    return textColorConfig;
};
var textColorPlugin = function (textColorConfig) { return ({
    name: 'textColor',
    marks: function () {
        return [{ name: 'textColor', mark: textColor }];
    },
    pmPlugins: function () {
        return [
            {
                name: 'textColor',
                plugin: function (_a) {
                    var dispatch = _a.dispatch;
                    return createPlugin(dispatch, pluginConfig(textColorConfig));
                },
            },
        ];
    },
    primaryToolbarComponent: function (_a) {
        var editorView = _a.editorView, popupsMountPoint = _a.popupsMountPoint, popupsBoundariesElement = _a.popupsBoundariesElement, popupsScrollableElement = _a.popupsScrollableElement, isToolbarReducedSpacing = _a.isToolbarReducedSpacing, dispatchAnalyticsEvent = _a.dispatchAnalyticsEvent, disabled = _a.disabled;
        var config = pluginConfig(textColorConfig);
        var showMoreColorsToggle = config && config.EXPERIMENTAL_allowMoreTextColors;
        return (React.createElement(WithPluginState, { plugins: {
                textColor: textColorPluginKey,
            }, render: function (_a) {
                var textColor = _a.textColor;
                return (React.createElement(ToolbarTextColor, { pluginState: textColor, isReducedSpacing: isToolbarReducedSpacing, editorView: editorView, popupsMountPoint: popupsMountPoint, popupsBoundariesElement: popupsBoundariesElement, popupsScrollableElement: popupsScrollableElement, showMoreColorsToggle: showMoreColorsToggle, dispatchAnalyticsEvent: dispatchAnalyticsEvent, disabled: disabled }));
            } }));
    },
}); };
export { textColorPluginKey };
export default textColorPlugin;
