// @atlaskit/avatar-group/dist/esm/components/AvatarGroup.js
import { __assign, __extends } from "tslib";
import React, { Component } from 'react';
import { ThemeProvider } from 'styled-components';
import DropdownMenu, { DropdownItemGroup } from '@atlaskit/dropdown-menu';
import Avatar from '@atlaskit/avatar';
import { Grid, Stack } from '../styled/AvatarGroup';
import MoreIndicator from './MoreIndicator';
import itemTheme from '../theme/itemTheme';
import AvatarGroupItem from './AvatarGroupItem';
var GROUP_COMPONENT = {
    grid: Grid,
    stack: Stack,
};
var MAX_COUNT = {
    grid: 11,
    stack: 5,
};
var AvatarGroup = /** @class */ (function (_super) {
    __extends(AvatarGroup, _super);
    function AvatarGroup() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AvatarGroup.prototype.getOverrides = function () {
        return {
            AvatarGroupItem: __assign({ render: function (Component, props, index) { return (React.createElement(Component, __assign({}, props, { key: index }))); } }, (this.props.overrides && this.props.overrides.AvatarGroupItem)),
            Avatar: __assign({ render: function (Component, props, index) { return (React.createElement(Component, __assign({}, props, { key: index }))); } }, (this.props.overrides && this.props.overrides.Avatar)),
        };
    };
    AvatarGroup.prototype.renderMoreDropdown = function (max, total) {
        var _this = this;
        var _a = this.props, appearance = _a.appearance, data = _a.data, borderColor = _a.borderColor, onMoreClick = _a.onMoreClick, showMoreButtonProps = _a.showMoreButtonProps, onAvatarClick = _a.onAvatarClick, size = _a.size, boundariesElement = _a.boundariesElement, testId = _a.testId;
        // bail if there's not enough items
        if (total <= max)
            return null;
        // prepare the button -- we'll use it twice
        var renderMoreButton = function (props) {
            if (props === void 0) { props = {}; }
            return (React.createElement(MoreIndicator, __assign({}, showMoreButtonProps, { borderColor: borderColor, count: total - max, isInteractive: true, isStack: appearance === 'stack', size: size, testId: testId && testId + "--overflow-menu--trigger" }, props)));
        };
        // bail if the consumer wants to handle onClick
        if (typeof onMoreClick === 'function') {
            return renderMoreButton({ onClick: onMoreClick });
        }
        // crop and prepare the dropdown items
        var items = data
            .slice(max)
            .map(function (avatar, index) {
            return _this.getOverrides().AvatarGroupItem.render(AvatarGroupItem, {
                avatar: avatar,
                onAvatarClick: onAvatarClick,
                testId: testId && testId + "--avatar-group-item-" + index,
            }, 
            // This index holds the true index,
            // adding up the index of non-overflowed avatars and overflowed avatars.
            index + max);
        });
        return (React.createElement(DropdownMenu, { trigger: renderMoreButton(), position: "bottom right", boundariesElement: boundariesElement, testId: testId && testId + "--overflow-menu", shouldFlip: true },
            React.createElement(ThemeProvider, { theme: itemTheme },
                React.createElement(DropdownItemGroup, null, items))));
    };
    AvatarGroup.prototype.render = function () {
        var _this = this;
        var _a = this.props, Item = _a.avatar, appearance = _a.appearance, borderColor = _a.borderColor, data = _a.data, maxCount = _a.maxCount, onAvatarClick = _a.onAvatarClick, size = _a.size, testId = _a.testId;
        // NOTE: conditionally defaulting the `maxCount` prop based on `appearance`
        var max = 
        // Check if it's zero since that's what it checked before, just in case.
        maxCount === undefined || maxCount === 0
            ? MAX_COUNT[appearance]
            : maxCount;
        var total = data.length;
        var Group = GROUP_COMPONENT[appearance];
        // Render (max - 1) avatars to leave space for moreIndicator
        var maxAvatar = total > max ? max - 1 : max;
        var items = data.slice(0, maxAvatar).map(function (avatar, idx) {
            return _this.getOverrides().Avatar.render(Item, __assign(__assign({}, avatar), { size: size,
                borderColor: borderColor, testId: testId && testId + "--avatar-" + idx, groupAppearance: appearance, onClick: avatar.onClick || onAvatarClick, stackIndex: max - idx }), idx);
        });
        return (React.createElement(Group, { size: size, "data-testid": testId && testId + "--avatar-group" },
            items,
            this.renderMoreDropdown(+maxAvatar, total)));
    };
    AvatarGroup.defaultProps = {
        appearance: 'stack',
        avatar: Avatar,
        showMoreButtonProps: {},
        size: 'medium',
    };
    return AvatarGroup;
}(Component));
export default AvatarGroup;
