// @atlaskit/reactions/dist/esm/containers/ReactionsContainer.js
import { __assign, __extends } from "tslib";
import { FabricElementsAnalyticsContext } from '@atlaskit/analytics-namespaced-context';
import { withAnalyticsEvents, } from '@atlaskit/analytics-next';
import React from 'react';
import { Reactions } from '../components/Reactions';
import { ReactionConsumer, } from '../reaction-store/ReactionConsumer';
import { ReactionStatus } from '../types/ReactionStatus';
var ReactionsContainerWithoutAnalytics = /** @class */ (function (_super) {
    __extends(ReactionsContainerWithoutAnalytics, _super);
    function ReactionsContainerWithoutAnalytics() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.renderChild = function (props) {
            var _a = _this.props, containerAri = _a.containerAri, ari = _a.ari;
            return (React.createElement(FabricElementsAnalyticsContext, { data: { containerAri: containerAri, ari: ari } },
                React.createElement(Reactions, __assign({ key: _this.props.containerAri + "|" + _this.props.ari }, _this.props, props))));
        };
        _this.stateMapper = function (state) {
            var _a = _this.props, containerAri = _a.containerAri, ari = _a.ari;
            var reactionsState = state && state.reactions[containerAri + "|" + ari];
            if (!state || !reactionsState) {
                return { status: ReactionStatus.notLoaded };
            }
            switch (reactionsState.status) {
                case ReactionStatus.ready:
                    return {
                        reactions: reactionsState.reactions,
                        status: reactionsState.status,
                        flash: state.flash[containerAri + "|" + ari],
                    };
                default:
                    return { status: ReactionStatus.loading };
            }
        };
        _this.actionsMapper = function (actions) { return ({
            loadReaction: function () {
                actions.getReactions(_this.props.containerAri, _this.props.ari);
            },
            onReactionClick: function (emojiId) {
                actions.toggleReaction(_this.props.containerAri, _this.props.ari, emojiId);
            },
            onReactionHover: function (emojiId) {
                actions.getDetailedReaction(_this.props.containerAri, _this.props.ari, emojiId);
            },
            onSelection: function (emojiId) {
                actions.addReaction(_this.props.containerAri, _this.props.ari, emojiId);
            },
        }); };
        return _this;
    }
    ReactionsContainerWithoutAnalytics.prototype.componentDidMount = function () {
        var _a = this.props, createAnalyticsEvent = _a.createAnalyticsEvent, store = _a.store;
        Promise.resolve(store).then(function (_store) {
            if (_store.setCreateAnalyticsEvent && createAnalyticsEvent) {
                _store.setCreateAnalyticsEvent(createAnalyticsEvent);
            }
        });
    };
    ReactionsContainerWithoutAnalytics.prototype.render = function () {
        return (React.createElement(ReactionConsumer, { store: this.props.store, actionsMapper: this.actionsMapper, stateMapper: this.stateMapper }, this.renderChild));
    };
    return ReactionsContainerWithoutAnalytics;
}(React.PureComponent));
var ReactionsContainer = withAnalyticsEvents()(ReactionsContainerWithoutAnalytics);
export default ReactionsContainer;
