// @atlaskit/router/dist/esm/controllers/router/index.js
import { __awaiter, __extends, __generator } from "tslib";
import React, { Component } from 'react';
import { DEFAULT_HISTORY } from '../../common/constants';
import { getRouterState, RouterContainer } from '../router-store';
/**
 * Default prop provider for the RouterContainer.
 *
 */
var Router = /** @class */ (function (_super) {
    __extends(Router, _super);
    function Router() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Ensures that the router store stops listening to history when the Router
     * is unmounted.
     */
    Router.prototype.componentWillUnmount = function () {
        if (!this.props.isStatic) {
            var state = getRouterState();
            state.unlisten && state.unlisten();
        }
    };
    Router.prototype.render = function () {
        var _a = this.props, children = _a.children, routes = _a.routes, history = _a.history, isStatic = _a.isStatic, transitionBlocker = _a.transitionBlocker, resourceContext = _a.resourceContext, resourceData = _a.resourceData;
        return (React.createElement(RouterContainer, { routes: routes, history: history, isStatic: isStatic, transitionBlocker: transitionBlocker, resourceContext: resourceContext, resourceData: resourceData, isGlobal: true }, children));
    };
    Router.defaultProps = {
        isStatic: false,
        history: DEFAULT_HISTORY,
        transitionBlocker: function () { return __awaiter(void 0, void 0, void 0, function () { return __generator(this, function (_a) {
            return [2 /*return*/, true];
        }); }); },
    };
    return Router;
}(Component));
export { Router };
