// @atlaskit/menu/dist/esm/components/section/styles.js
import { __assign } from "tslib";
import { N30 } from '@atlaskit/theme/colors';
export var menuGroupCSS = function (_a) {
    var minWidth = _a.minWidth, maxWidth = _a.maxWidth, minHeight = _a.minHeight, maxHeight = _a.maxHeight;
    return ({
        display: 'flex',
        flexDirection: 'column',
        overflow: 'auto',
        maxWidth: maxWidth,
        minWidth: minWidth,
        maxHeight: maxHeight,
        minHeight: minHeight,
    });
};
export var sectionCSS = function (isScrollable, hasSeparator) {
    return __assign(__assign(__assign({}, (isScrollable
        ? {
            flexShrink: 1,
            overflow: 'auto',
        }
        : { flexShrink: 0 })), (hasSeparator
        ? {
            borderTop: "2px solid " + N30,
        }
        : {
            // this is to ensure that adjacent sections without separators don't get additional margins.
            '[data-section] + &': {
                marginTop: -6,
            },
        })), { '&:focus': {
            // NOTE: Firefox allows elements that have "overflow: auto" to gain focus (as if it had tab-index="0")
            // We have made a deliberate choice to leave this behaviour as is.
            // This makes the outline go inside by 1px so it can actually be displayed
            // else it gets cut off from the overflow: scroll from the parent menu group.
            outlineOffset: -1,
        } });
};
