// @atlassian/connect-module-core/dist/es6/modules/flag/FlagSimpleXdmDefinitionBuilder.js
var FlagSimpleXdmDefinitionBuilder = /** @class */ (function () {
    function FlagSimpleXdmDefinitionBuilder(flagProvider) {
        if (!flagProvider) {
            throw new Error('flagProvider is not defined');
        }
        this.flagProvider = flagProvider;
    }
    FlagSimpleXdmDefinitionBuilder.prototype.buildSimpleXdmDefinition = function () {
        var flagProvider = this.flagProvider;
        /*
         * This class is a proxy to the flag object created by the flag provider. It is
         * equivalent to the Flag class defined in atlassian-connect-js/src/modules/flag.js
         * except that it does not pull in any AUI flag implementation.
         */
        var FlagProxy = /** @class */ (function () {
            function FlagProxy(options, callback) {
                options.id = callback._id;
                var flagId = callback._id;
                var actions = [];
                if (typeof options.actions === 'object') {
                    actions = Object.getOwnPropertyNames(options.actions).map(function (key) {
                        return {
                            actionKey: key,
                            actionText: options.actions[key],
                            executeAction: function () {
                                console.log('This needs to be implemented by broadcasting a "flag-action-invoked" event.');
                            }
                        };
                    });
                }
                var flagOptions = {
                    id: flagId,
                    title: options.title,
                    body: options.body,
                    actions: actions,
                    onClose: function () {
                        console.log('This needs to be implemented by broadcasting a "flag.close" event.');
                    },
                    close: options.close,
                    type: options.type.toLowerCase()
                };
                this.flag = flagProvider.create(flagOptions);
            }
            FlagProxy.prototype.close = function () {
                this.flag.close();
            };
            return FlagProxy;
        }());
        return {
            create: {
                constructor: FlagProxy,
                close: FlagProxy.prototype.close
            }
        };
    };
    return FlagSimpleXdmDefinitionBuilder;
}());
export { FlagSimpleXdmDefinitionBuilder };
