// @atlassian/connect-module-core/dist/es6/modules/inlinedialog/InlineDialogModule.js
import * as tslib_1 from "tslib";
import { BaseModule } from '../base/Module';
import { InlineDialogSimpleXdmDefinitionBuilder } from './InlineDialogSimpleXdmDefinitionBuilder';
export var InlineDialogModuleDefinitions;
(function (InlineDialogModuleDefinitions) {
    InlineDialogModuleDefinitions.INLINE_DIALOG_MODULE_NAME = 'inlineDialog';
    var Props = /** @class */ (function () {
        function Props(adaptor, inlineDialogProvider) {
            this.adaptor = adaptor;
            this.inlineDialogProvider = inlineDialogProvider;
        }
        return Props;
    }());
    InlineDialogModuleDefinitions.Props = Props;
    var State = /** @class */ (function () {
        function State(inlineDialogProvider) {
            this.enabled = true;
            this.registered = false;
            this.inlineDialogProvider = inlineDialogProvider;
            this.simpleXdmDefinitionBuilder = new InlineDialogSimpleXdmDefinitionBuilder(inlineDialogProvider);
        }
        return State;
    }());
    InlineDialogModuleDefinitions.State = State;
})(InlineDialogModuleDefinitions || (InlineDialogModuleDefinitions = {}));
/**
 * This module provides inlineDialog related functionality.
 */
var InlineDialogModule = /** @class */ (function (_super) {
    tslib_1.__extends(InlineDialogModule, _super);
    function InlineDialogModule(props) {
        return _super.call(this, props) || this;
    }
    InlineDialogModule.prototype.componentWillMount = function () {
        if (!this.props.inlineDialogProvider) {
            throw new Error('this.props.inlineDialogProvider is not defined');
        }
        this.setState(new InlineDialogModuleDefinitions.State(this.props.inlineDialogProvider));
    };
    InlineDialogModule.prototype.getModuleRegistrationName = function () {
        return InlineDialogModuleDefinitions.INLINE_DIALOG_MODULE_NAME;
    };
    InlineDialogModule.prototype.getSimpleXdmDefinitionBuilder = function () {
        return this.state.simpleXdmDefinitionBuilder;
    };
    InlineDialogModule.prototype.getProvider = function () {
        return this.props.inlineDialogProvider;
    };
    InlineDialogModule.prototype.render = function () {
        return null;
    };
    return InlineDialogModule;
}(BaseModule));
export { InlineDialogModule };
