// @atlaskit/editor-core/dist/esm/plugins/expand/toolbar.js
import RemoveIcon from '@atlaskit/icon/glyph/editor/remove';
import commonMessages from '../../messages';
import { deleteExpand } from './commands';
import { getPluginState } from './pm-plugins/main';
import { hoverDecoration } from '../base/pm-plugins/decoration';
export var getToolbarConfig = function (state, _a) {
    var formatMessage = _a.formatMessage;
    var expandRef = getPluginState(state).expandRef;
    if (expandRef) {
        var _b = state.schema.nodes, nestedExpand = _b.nestedExpand, expand = _b.expand;
        return {
            title: 'Expand toolbar',
            getDomRef: function () { return expandRef; },
            nodeType: [nestedExpand, expand],
            offset: [0, 6],
            items: [
                {
                    type: 'button',
                    appearance: 'danger',
                    icon: RemoveIcon,
                    onClick: deleteExpand(),
                    onMouseEnter: hoverDecoration([nestedExpand, expand], true),
                    onMouseLeave: hoverDecoration([nestedExpand, expand], false),
                    title: formatMessage(commonMessages.remove),
                },
            ],
        };
    }
    return;
};
