// @atlassian/connect-module-core/dist/es6/modules/dropdown/DropdownModule.js
import * as tslib_1 from "tslib";
import { BaseModule } from '../base/Module';
import { DropdownSimpleXdmDefinitionBuilder } from './DropdownSimpleXdmDefinitionBuilder';
export var DropdownModuleDefinitions;
(function (DropdownModuleDefinitions) {
    DropdownModuleDefinitions.DROPDOWN_MODULE_NAME = 'dropdown';
    var Props = /** @class */ (function () {
        function Props(adaptor, dropdownProvider) {
            this.adaptor = adaptor;
            this.dropdownProvider = dropdownProvider;
        }
        return Props;
    }());
    DropdownModuleDefinitions.Props = Props;
    var State = /** @class */ (function () {
        function State(dropdownProvider) {
            this.enabled = true;
            this.registered = false;
            this.dropdownProvider = dropdownProvider;
            this.simpleXdmDefinitionBuilder = new DropdownSimpleXdmDefinitionBuilder(dropdownProvider);
        }
        return State;
    }());
    DropdownModuleDefinitions.State = State;
})(DropdownModuleDefinitions || (DropdownModuleDefinitions = {}));
/**
 * This module provides dropdown related functionality.
 */
var DropdownModule = /** @class */ (function (_super) {
    tslib_1.__extends(DropdownModule, _super);
    function DropdownModule(props) {
        return _super.call(this, props) || this;
    }
    DropdownModule.prototype.componentWillMount = function () {
        if (!this.props.dropdownProvider) {
            throw new Error('this.props.dropdownProvider is not defined');
        }
        this.setState(new DropdownModuleDefinitions.State(this.props.dropdownProvider));
    };
    DropdownModule.prototype.getModuleRegistrationName = function () {
        return DropdownModuleDefinitions.DROPDOWN_MODULE_NAME;
    };
    DropdownModule.prototype.getSimpleXdmDefinitionBuilder = function () {
        return this.state.simpleXdmDefinitionBuilder;
    };
    DropdownModule.prototype.getProvider = function () {
        return this.props.dropdownProvider;
    };
    DropdownModule.prototype.render = function () {
        return null;
    };
    return DropdownModule;
}(BaseModule));
export { DropdownModule };
