// @atlaskit/editor-core/dist/esm/ui/ColorPalette/Color/index.js
import { __extends } from "tslib";
import React from 'react';
import { PureComponent } from 'react';
import { defineMessages, injectIntl } from 'react-intl';
import EditorDoneIcon from '@atlaskit/icon/glyph/editor/done';
import { colors } from '@atlaskit/theme';
import { Button, ButtonWrapper } from './styles';
import Tooltip from '@atlaskit/tooltip';
// IMO these should live inside @atlaskit/theme
var messages = defineMessages({
    selected: {
        id: 'fabric.editor.selected',
        defaultMessage: 'Selected',
        description: 'If the item is selected or not.',
    },
});
var Color = /** @class */ (function (_super) {
    __extends(Color, _super);
    function Color() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.onMouseDown = function (e) {
            e.preventDefault();
        };
        _this.onClick = function (e) {
            var _a = _this.props, onClick = _a.onClick, value = _a.value;
            e.preventDefault();
            onClick(value);
        };
        return _this;
    }
    Color.prototype.render = function () {
        var _a = this.props, tabIndex = _a.tabIndex, value = _a.value, label = _a.label, isSelected = _a.isSelected, borderColor = _a.borderColor, _b = _a.checkMarkColor, checkMarkColor = _b === void 0 ? colors.N0 : _b, formatMessage = _a.intl.formatMessage;
        return (React.createElement(Tooltip, { content: label },
            React.createElement(ButtonWrapper, null,
                React.createElement(Button, { onClick: this.onClick, onMouseDown: this.onMouseDown, tabIndex: tabIndex, className: "" + (isSelected ? 'selected' : ''), style: {
                        backgroundColor: value || 'transparent',
                        border: "1px solid " + borderColor,
                    } }, isSelected && (React.createElement(EditorDoneIcon, { primaryColor: checkMarkColor, label: formatMessage(messages.selected) }))))));
    };
    return Color;
}(PureComponent));
export default injectIntl(Color);
